dart/dynamics/HierarchicalIK.cpp

```diff
@@ -0,0 +1,809 @@
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dart/dynamics/DegreeOfFreedom.h"
+#include "dart/dynamics/HierarchicalIK.h"
+#include "dart/dynamics/BodyNode.h"
+#include "dart/dynamics/EndEffector.h"
+#include "dart/dynamics/Skeleton.h"
+
+namespace dart {
+namespace dynamics {
+
+//==============================================================================
+bool HierarchicalIK::solve(bool _resetConfiguration)
+{
+  if(nullptr == mSolver)
+  {
+    dtwarn << "[HierarchicalIK::solve] The Solver for a HierarchicalIK module "
+           << "associated with [" << mSkeleton.lock()->getName() << "] is a "
+           << "nullptr. You must reset the module's Solver before you can use "
+           << "it.\n";
+    return false;
+  }
+
+  if(nullptr == mProblem)
+  {
+    dtwarn << "[HierarchicalIK::solve] The Problem for a HierarchicalIK module "
+           << "associated with [" << mSkeleton.lock()->getName() << "] is a "
+           << "nullptr. You must reset the module's Problem before you can use "
+           << "it.\n";
+    return false;
+  }
+
+  const SkeletonPtr& skel = getSkeleton();
+
+  if(nullptr == skel)
+  {
+    dtwarn << "[HierarchicalIK::solve] Calling a HierarchicalIK module which "
+           << "is associated with a Skeleton that no longer exists.\n";
+    return false;
+  }
+
+  const size_t nDofs = skel->getNumDofs();
+  mProblem->setDimension(nDofs);
+
+  mProblem->setInitialGuess(skel->getPositions());
+
+  Eigen::VectorXd bounds(nDofs);
+  for(size_t i=0; i < nDofs; ++i)
+    bounds[i] = skel->getDof(i)->getPositionLowerLimit();
+  mProblem->setLowerBounds(bounds);
+
+  for(size_t i=0; i < nDofs; ++i)
+    bounds[i] = skel->getDof(i)->getPositionUpperLimit();
+  mProblem->setUpperBounds(bounds);
+
+  refreshIKHierarchy();
+
+  if(!_resetConfiguration)
+    return mSolver->solve();
```

Comment:
In general, I find it useful to add an explanatory comment in situations where we are working around an issue in an upstream library, since future maintainers will have no idea what it is doing if the upstream issue gets resolved (and be able to remove/de-cruft it).